[Merged by Bors] - feat(group_theory/perm): define the permutation `(0 1 ... (i : fin n))`

Vierkantor opened the pull request:
I tried a number of definitions and it looks like this is the least awkward one to prove with. In any case, using the API should allow you to avoid unfolding the definition.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->
- [x] depends on: #6959
- [x] depends on: #7043 
[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Vierkantor committed:
feat(group_theory/perm): define the permutation `(0 1 ... (i : fin n))`

I tried a number of definitions and it looks like this is the least awkward one
to prove with. In any case, using the API should allow you to avoid unfolding
the definition.

Vierkantor commented on src/group_theory/perm/fin.lean:
```diff
@@ -62,3 +62,121 @@ lemma finset.univ_perm_fin_succ {n : ℕ} :
+      contradiction },
+    { intro hji,
+      have hji := congr_arg (coe : fin _ → ℕ) hji,
+      exact h.ne hji } },
```
I don't like how this part of the proof looks. It's repetitive, but not exactly identical. Anyone have a nice idea to golf this?

Vierkantor committed:
Define `cycle_range_of_apply`

eric-wieser commented on src/group_theory/perm/fin.lean:
```diff
@@ -62,3 +62,130 @@ lemma finset.univ_perm_fin_succ {n : ℕ} :
+-- We could write this as a product over `list.range`,
+-- but the explicit `mk`s would become `coe`s and make the proofs a lot more annoying.
+-- We could also write this as two `if`s, but it is very annoying to prove that
+-- indeed results in a permutation.
```
Does it make sense to break this definition into two pieces; one that does the rotation, and one that does the pass-through?
```lean
/-- A permutation of `fin m` that rotates elements by `n` -/
def cycle (n m : ℕ) : perm (fin m) :=
{ to_fun := λ i, i + n,  -- this would work for zmod...
  inv_fun := λ i, i - n,
  left_inv := λ i, sorry,
  right_inv := λ i, sorry, }

/-- rotate the first `m` elements by `1`, leave the remaining `n` unchanged. -/
def cycle_range (m : ℕ) (n : ℕ) : perm (fin $ m + n) :=
sum_fin_sum_equiv.perm_congr ((cycle 1 m).sum_congr (equiv.refl $ fin n))
```

Vierkantor replied:
Hmm, I'm not really convinced that those are the natural ways to carve up this definition. In fact, wouldn't it be nicer to define `cycle n m` as `(cycle_range (fin.last n)) ^ m`? (Also, `cycle` is probably not the right name for that map since [`is_cycle`](https://leanprover-community.github.io/mathlib_docs/group_theory/perm/cycles.html#equiv.perm.is_cycle) has a specific meaning in group theory. If you want it, it should be something like `add_perm`.)

If we were to generalize, group theorists would prefer `cycle (xs : list α) : perm α`, sending `xs.nth 0` to `xs.nth 1`, `xs.nth 1` to `xs.nt 2`, ... `xs.nth (length xs - 1)` to `xs.nth 0`. (And `fin.cycle_range i` would just be `perm.cycle (list.range i)`.) But list indices are not so nice to work with, so you'd want to use `fin` indices instead, and the result is `cycle_range` composed with some injective `fin n -> α`.

eric-wieser replied:
I guess what I'm finding weird here is the "cycle the first m, leave the rest" behavior. We have lots of combinators for "permute just these elements", what we're missing is the "cycle all elements" operation.

> But list indices are not so nice to work with, so you'd want to use fin indices instead, and the result is cycle_range composed with some injective fin n -> α.

Here, the "permute just these elements" combinator would be:
```lean
-- I'm surprised this is missing - added in #6817
instance decidable_fintype_range {α ι : Type*} [fintype ι] (f : ι → α) [decidable_eq α] :
  decidable_pred (λ (x : α), x ∈ set.range f) :=
λ x, begin
  unfold set.range,
  dsimp [set.mem_set_of_eq],
  apply_instance,
end

-- computable version of `equiv.set.range` - a better version PR'd as #6821
def function.embedding.equiv_range {α ι : Type*} [fintype ι] [decidable_eq α] (f : ι ↪ α) :
  ι ≃ set.range f :=
{ to_fun := λ x, ⟨f x, x, rfl⟩,
  inv_fun := f.inv_of_mem_range,
  left_inv := f.right_inv_of_inv_of_mem_range,
  right_inv := λ x, subtype.eq $ f.left_inv_of_inv_of_mem_range x }

def function.embedding.range_perm {α : Type*} {n : ℕ} (f : fin n ↪ α) [decidable_eq α] (e : perm $ fin n) : perm α :=
(f.equiv_range.perm_congr e).subtype_congr (equiv.refl _)
-- reads: "apply e on the range of `f`, and apply the identity elsewhere
```

Vierkantor replied:
Aha, so your goal looks more like using an embedding to turn  "permute the whole of `fin n`" into "permute the `fin n` subset of `fin k`"?

eric-wieser replied:
Yes, with the bonus that now we're not really tied to `fin` at all - the only thing that actually requires `fin` is the cycle sending `i → i + 1`, `fin.last → 0`

Vierkantor replied:
[I managed to define](https://github.com/leanprover-community/mathlib/commit/0107630b0d4b92f424ad25f8d381a5f614b15b12#diff-61fef3e3002e0edd652a7dae93e737aa1111caf3ebb9a05cb007417f73837ceaR79) `fin.cycle_range` in terms of a new `fin.cycle_all`, but I'm really not convinced this is the way to go. (I would much rather define `fin.cycle_all` in terms of `fin.cycle_range`.) Trying to compute with these `subtype_congr` and `perm_congr` means 10s of lines of pushing around trivial inequalities, and it's really hard to get the sign of these permutations without basically resorting to the original definition of `cycle_range` anyway. Or is there some golfing trick that I'm missing?

Vierkantor replied:
[I just pushed](https://github.com/leanprover-community/mathlib/pull/6815/commits/aaecafe875f61f848a2578aaa7d6c5571a0214b0) a definition of `cycle_all` and the same results, defined in terms of `cycle_range`. I definitely know which commit I prefer :)

eric-wieser replied:
Here's a much shorter definition of `cycle_all`:
```lean
-- import data.zmod.basic for fin.comm_ring

/-- `fin.cycle_all` is the cycle `(0 1 2 ... fin.last n)`. -/
def cycle_all : Π (n : ℕ), perm (fin n)
| 0 := 1
| (nat.succ n) := let inst := fin.comm_ring n in by exactI equiv.add_right 1
```

eric-wieser replied:
I wasn't able to try out your branch because of build failures in `equiv/basic`

eric-wieser replied:
To make the `equiv.add_right` definition non-painful, we need #6848

Vierkantor replied:
> ```lean
> -- import data.zmod.basic for fin.comm_ring
> 
> /-- `fin.cycle_all` is the cycle `(0 1 2 ... fin.last n)`. -/
> def cycle_all : Π (n : ℕ), perm (fin n)
> | 0 := 1
> | (nat.succ n) := let inst := fin.comm_ring n in by exactI equiv.add_right 1
> ```

So we standardised `fin` back on wrapping addition? That should help a bunch indeed!

Vierkantor replied:
I'll see in a couple hours if I can get it to work.

eric-wieser replied:
I think you can get `sign (cycle_all n)` by showing `cycle_all n.succ` is equal to `decompose_fin.symm (1, cycle_all n)` or similar.

Vierkantor replied:
Took a bit longer than expected, the proofs are still not as nice as I was hoping, but I can live with this approach.

Vierkantor committed:
Generalize `sign_cycle_range`

Vierkantor committed:
Lint fix

Vierkantor committed:
For comparison: the same definition and results for `cycle_all`, defined in terms of `cycle_range`

Vierkantor committed:
Define `cycle_range` in terms of `cycle_all`

I'm still not convinced that this really is nicer than defining `cycle_all` in
terms of `cycle_range`, but it's less bad than the previous attempt.

Vierkantor committed:
Cleanup, move definitions to the appropriate locations

eric-wieser commented on src/data/fin.lean:
```diff
@@ -345,6 +349,24 @@ lemma val_add {n : ℕ} : ∀ a b : fin n, (a + b).val = (a.val + b.val) % n
+  cases n,
+  { apply subsingleton.elim },
+  { ext, rw [coe_add, coe_zero, coe_last, coe_one, nat.mod_self] }
+end
```
I had this proof locally as:
```suggestion
@[simp] lemma last_add_one : ∀ {n : ℕ}, fin.last n + 1 = 0
| 0 := subsingleton.elim _ _
| (n + 1) := subtype.eq $ by simp [fin.coe_add, fin.coe_one]
```
but I guess yours avoids `simp`

eric-wieser replied:
I think anothernice simp lemma to have would be
```
lemma fin.zero_sub_one : ∀ {n : ℕ}, 0 - 1 = fin.last n
| 0 := subsingleton.elim _ _
| (n + 1) := subtype.eq $ by {
  unfold has_zero.zero has_one.one fin.of_nat has_sub.sub,
  unfold fin.sub fin.last,
  simp [fin.sub, fin.coe_add, fin.coe_one],
  sorry  -- I didn't get as far as proving this, as `has_sub (fin n)` didn't match `fin.sub` until recently.
}
```

eric-wieser commented on src/group_theory/perm/fin.lean:
```diff
@@ -62,3 +63,158 @@ lemma finset.univ_perm_fin_succ {n : ℕ} :
+def cycle_range {n : ℕ} (i : fin n) : perm (fin n) :=
+((equiv.set.range_of_left_inverse' (fin.cast_le i.2) coe (by { intros x, ext, simp }))
+  .perm_congr (cycle_all (i + 1)))
+  .subtype_congr 1
```
I was really expecting you to need the unprimed version here! What type does `coe` have?

Vierkantor replied:
IIRC `coe : fin (↑i + 1) → ℕ` (which is why `fin.cast_le i.2` works, `i.2 : ↑i < n`, which is defeq to `↑i + 1 ≤ n`.

eric-wieser replied:
Oh, I see my mistake; I thought `fin_cycle_range i` cycled the first `i` elements, not the first `i + 1`.

Vierkantor replied:
Indeed, it's the cycle including `i`. This way, `cycle_range 0` permutes nothing and `cycle_range (last n)` permutes everything.

Vierkantor replied:
That reminds me, I should add a `simp` lemma `cycle_range (last n) = cycle_all (n + 1)`.

eric-wieser commented on src/group_theory/perm/fin.lean:
```diff
@@ -62,3 +63,158 @@ lemma finset.univ_perm_fin_succ {n : ℕ} :
+/-- `fin.cycle_all` is the cycle `(0 1 2 ... fin.last n)`.
+
+If `n` is `0` or `1`, this will be the identity permutation,
+otherwise it sends `i` to `i + 1` and `fin.last n` to `0`.
```
```suggestion
otherwise it sends `i` to `i + 1` and `fin.last (n - 1)` to `0`.
```

eric-wieser commented on src/group_theory/perm/fin.lean:
```diff
@@ -62,3 +63,158 @@ lemma finset.univ_perm_fin_succ {n : ℕ} :
+  { rw cycle_all_succ, simp [ih, pow_succ] },
+end
+
+/-- `fin.cycle_range i` is the cycle `(0 1 2 ... i)`. -/
```
```suggestion
/-- `fin.cycle_range i` is the cycle `(0 1 2 ... i)` leaving `(i+1 ... (n-1))` unchanged. -/
```

eric-wieser commented on src/group_theory/perm/fin.lean:
```diff
@@ -62,3 +63,158 @@ lemma finset.univ_perm_fin_succ {n : ℕ} :
+  refine fin.cases _ (λ i, _) i,
+  { simp },
+  rw [coe_cycle_all, decompose_fin_symm_apply_succ,
+      if_congr (show i.succ = last (n + 1) ↔ i = last n, from _) rfl rfl],
```
I think this `show` might be a good lemma to extract, `succ_eq_last_succ`

Vierkantor replied:
It seems a rather situational lemma, but I'm happy to extract it :)

eric-wieser replied:
It's only worth extracting if you also use the new lemma here!

Vierkantor committed:
Fix duplicate namespace

Vierkantor committed:
Synthesize @eric-wieser's proof and mine

Vierkantor committed:
Define `succ_eq_last_succ`

eric-wieser commented on src/data/equiv/basic.lean:
```diff
@@ -1688,11 +1688,22 @@ def range_of_left_inverse {α β : Sort*}
 
+@[simp] lemma range_of_left_inverse_symm_apply {α β : Sort*}
+  (f : α → β) (f_inv hf) (x : set.range f) :
+  (equiv.set.range_of_left_inverse f f_inv hf).symm x = f_inv ⟨classical.some x.2⟩ x :=
```
Any reason not to let `simp` generate this one, which avoids using `classical`?

eric-wieser replied:
Oh, I guess I should have implemented `range_of_left_inverse` using `nonempty_of_exists`:
```suggestion
  (equiv.set.range_of_left_inverse f f_inv hf).symm x = f_inv (nonempty_of_exists x.2) x :=
```

Vierkantor replied:
~~`simps` gives a RHS containing `range_of_left_inverse._proof_2`, which doesn't seem so nice to me. Or should I put an explicit `exists.elim` in the definition of `range_of_left_inverse`?~~

Vierkantor committed:
Clean up definition `equiv.set.range_of_left_inverse` so we can `@[simps]` it

Vierkantor committed:
Add `cycle_range_last` (and helper lemmas)

Vierkantor committed:
Clarify docstrings

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

Vierkantor committed:
Actually use the extracted lemma

eric-wieser commented on src/data/fin.lean:
```diff
@@ -504,6 +522,17 @@ order_embedding.of_strict_mono (λ a, cast_lt a (lt_of_lt_of_le a.2 h)) $ λ a b
+set.ext (λ x, ⟨λ ⟨y, hy⟩, hy ▸ y.2, λ hx, ⟨⟨x, hx⟩, fin.ext rfl⟩⟩)
+
+@[simp] lemma coe_set_range_cast_le_symm {n k : ℕ} (h : n ≤ k) (i : fin k) (hi) :
+  ((equiv.set.range _ ((cast_le h).injective)).symm ⟨i, hi⟩ : ℕ) = i :=
```
nit: parens
```suggestion
  ((equiv.set.range _ (cast_le h).injective).symm ⟨i, hi⟩ : ℕ) = i :=
```

Do you actually use this lemma anywhere?

Vierkantor replied:
I don't think so (nor the one for `cast_succ`), since everything is now `range_of_left_inverse`. Should I just get rid of them because they are too specific?

eric-wieser replied:
I'd probably be inclined to get rid of at least the `equiv.set.range` lemmas, yes - if `coe` is the inverse, then probably `equiv.set.range` was the wrong definition to use in the first place.

The ones about `set.range` are probably harmless to keep,

eric-wieser commented on src/data/fin.lean:
```diff
@@ -196,6 +196,10 @@ def coe_embedding (n) : (fin n) ↪o ℕ :=
 
+/-- Use the ordering on `fin n` for checking recursive definitions. -/
+instance {n : ℕ} : has_well_founded (fin n) :=
+⟨_, measure_wf coe⟩
```
You no longer use this in this PR, right? Would you mind adding an example of what this enables in its docstring?

Vierkantor committed:
delete unecessary parens

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

Vierkantor committed:
Give an example of the `has_well_founded (fin n)` instance

pechersky commented on src/group_theory/perm/fin.lean:
```diff
@@ -62,3 +63,170 @@ lemma finset.univ_perm_fin_succ {n : ℕ} :
+If `n` is `0` or `1`, this will be the identity permutation,
+otherwise it sends `i` to `i + 1` and `fin.last (n - 1)` to `0`.
+-/
+def cycle_all : Π (n : ℕ), perm (fin n)
```
Isn't this 
```
/-- Rotate `fin n` one step to the right. -/
def fin_rotate : Π n, fin n ≃ fin n
| 0 := equiv.refl _
| (n+1) := fin_add_flip.trans (fin_congr (add_comm _ _))
```
defined in `data.equiv.fin`?

eric-wieser replied:
Nice find. It would be good to prove:
```lean
section
local attribute [instance] fin.comm_ring
lemma fin_rotate_succ_eq_add_right {n : nat} :
  fin.rotate n.succ = equiv.add_right 1 := sorry
end
```

pechersky replied:
I'm currently testing defining `cycle_range` just by composing `fin.rotate` with `sum_fin_sum_equiv`.

Vierkantor replied:
Indeed, it is the same. It had so little API that the only reduction in lines was not having to import `data.zmod.basic`.

Vierkantor committed:
`cycle_all` already existed, but not any of the API

pechersky commented on src/data/equiv/fin.lean:
```diff
@@ -227,6 +227,47 @@ begin
+@[simp] lemma fin_rotate_apply_zero {n : ℕ} : fin_rotate n.succ 0 = 1 :=
+by rw [fin_rotate_succ_apply, zero_add]
+
+@[simp] lemma fin_rotate_apply_last {n : ℕ} : fin_rotate n.succ (fin.last n) = 0 :=
```
This is `fin_rotate_last` above.

pechersky commented on src/data/equiv/fin.lean:
```diff
@@ -227,6 +227,47 @@ begin
+  rw [fin_rotate, equiv.trans_apply, fin_add_flip_apply_left i_lt _],
+  ext,
+  simp [add_comm 1 i, fin.coe_add_one coe_i_lt]
+end
```
```suggestion
lemma fin_rotate_succ_apply {n : ℕ} (i : fin n.succ) :
  fin_rotate n.succ i = i + 1 :=
begin
  cases n,
  { simp },
  rcases i.le_last.eq_or_lt with rfl|h,
  { simp },
  { cases i,
    simp only [fin.lt_iff_coe_lt_coe, fin.coe_last, fin.coe_mk] at h,
    simp [fin_rotate_of_lt h, fin.eq_iff_veq, fin.add_def, nat.mod_eq_of_lt (nat.succ_lt_succ h)] },
end
```

Vierkantor commented on src/data/equiv/fin.lean:
```diff
@@ -227,6 +227,47 @@ begin
+
+@[simp] lemma fin_rotate_apply_last {n : ℕ} : fin_rotate n.succ (fin.last n) = 0 :=
+by { ext, rw [fin_rotate_succ_apply, fin.last_add_one] }
+
```
```suggestion

```

Vierkantor committed:
Apply suggestions from code review

Co-authored-by: Yakov Pechersky <pechersky@users.noreply.github.com>

pechersky committed:
rephrase definitions, which simplifies proofs

TODO: move new defns and lemmas to appropriate files

pechersky committed:
Merge branch 'fin_cycle_range' of https://github.com/leanprover-community/mathlib into fin_cycle_range

Vierkantor committed:
Move the `to_equiv_range` definitions to a new file

This seemed like the easiest way to get out of the import tangle: add a new
node.

Also a few docstrings and other lint fixes.

pechersky committed:
rename lemmas and expand docstrings

pechersky committed:
equality lemma to noncomputable

pechersky committed:
revert back to more efficient def

pechersky committed:
feat(data/equiv/fintype): computable bijections and perms from/to fintype

Using exhaustive search, we can upgrade embeddings from fintypes into
equivs, and transport perms across embeddings. The computational
performance is poor, so the docstring suggests alternatives when an
explicit inverse is known.

pechersky committed:
Merge branch 'pechersky/fintype-computable-equiv' into fin_cycle_range

pechersky committed:
Merge branch 'master' into fin_cycle_range

pechersky committed:
fix renames

pechersky committed:
finish renames

eric-wieser commented on src/data/equiv/basic.lean:
```diff
@@ -1750,6 +1750,11 @@ lemma of_left_inverse_eq_of_injective {α β : Type*} [nonempty α]
+  (f : α → β) (f_inv : β → α) (hf : left_inverse f_inv f) :
+  of_left_inverse' f f_inv hf = of_injective f hf.injective :=
+of_left_inverse_eq_of_injective _ _ _
+
```
I'm pretty sure that `[nonempty α]` isn't needed to prove either this lemma or the one above

eric-wieser commented on src/data/equiv/fin.lean:
```diff
@@ -244,6 +244,42 @@ begin
+  split_ifs with h,
+  { simp [fin_rotate_succ_apply, h] },
+  exact coe_fin_rotate_of_ne_last h
+end
```
Does it make more sense to state this as
```suggestion
lemma coe_add_one {n : ℕ} (i : fin n.succ) :
  (i + 1 : ℕ) = if i = fin.last n then 0 else i + 1 :=
```
and then put it in `src/data/fin`?

eric-wieser replied:
Now that you've done that, can you prove this lemma using the new `coe_add_one`?

eric-wieser commented on src/data/equiv/fin.lean:
```diff
@@ -244,6 +244,42 @@ begin
+@[simp] lemma fin_rotate_apply_zero {n : ℕ} : fin_rotate n.succ 0 = 1 :=
+by rw [fin_rotate_succ_apply, zero_add]
+
+lemma coe_fin_rotate_of_ne_last {n : ℕ} {i : fin n.succ} (h : i ≠ fin.last n) :
+  (fin_rotate n.succ i : ℕ) = i + 1 :=
```
I think this lemma belongs next to `coe_add_one`, perhaps as:
```suggestion
lemma coe_add_one_of_ne_last {n : ℕ} {i : fin n.succ} (h : i ≠ fin.last n) :
  (i + 1 : ℕ) = i + 1 :=
```

eric-wieser commented on src/data/equiv/fin.lean:
```diff
@@ -244,6 +244,42 @@ begin
+@[simp] lemma fin_rotate_one : fin_rotate 1 = equiv.refl _ :=
+subsingleton.elim _ _
+
+lemma fin_rotate_succ_apply {n : ℕ} (i : fin n.succ) :
```
```suggestion
@[simp] lemma fin_rotate_succ_apply {n : ℕ} (i : fin n.succ) :
```

pechersky committed:
feat(logic/basic): singleton_of_not_nonempty

pechersky committed:
incorporate suggestions

Vierkantor committed:
Finish the `coe_add_one` -> `coe_add_one_of_lt` rename

Vierkantor committed:
Use `coe_fin_rotate` in the proof of `cycle_range_of_le`

Vierkantor committed:
Merge remote-tracking branch 'mathlib/master' into fin_cycle_range

Vierkantor committed:
Use `fin.coe_add_one` in the proof of `coe_fin_rotate`

Vierkantor committed:
Lint fix

jcommelin reviewed (approved):
Thanks :tada:

bors merge
